linc_convert/utils/zarr/generate_pyramid.py

```diff
@@ -0,0 +1,166 @@
+        Zarr group object
+    levels : int
+        Number of additional levels to generate.
+        By default, stop when all dimensions are smaller than their
+        corresponding chunk size.
+    ndim : int
+        Number of spatial dimensions.
+    max_load : int
+        Maximum number of voxels to load along each dimension.
+    mode : {"mean", "median"}
+        Whether to use a mean or median moving window.
+    no_pyramid_axis : int | None
+        Axis that should not be downsampled. If None, downsample
+        across all three dimensions.
+    Returns
+    -------
+    shapes : list[list[int]]
+        Shapes of all levels, from finest to coarsest, including the
+        existing top level.
+    """
+    if max_load % 2:
+        max_load += 1
+
+    # Read properties from base level
+    shape = list(omz["0"].shape)
+    chunk_size = omz["0"].chunks
+    opt = {
+        "dimension_separator": omz["0"]._dimension_separator,
+        "order": omz["0"]._order,
+        "dtype": omz["0"]._dtype,
+        "fill_value": omz["0"]._fill_value,
+        "compressor": omz["0"]._compressor,
+        "chunks": omz["0"].chunks,
+    }
+
+    # Select windowing function
+    if mode == "median":
+        func = np.median
+    else:
+        assert mode == "mean"
+        func = np.mean
+
+    level = 0
+    batch, shape = shape[:-ndim], shape[-ndim:]
+    allshapes = [shape]
+
+    while True:
+        level += 1
+
+        # Compute downsampled shape
+        prev_shape, shape = shape, []
+        for i, length in enumerate(prev_shape):
+            if i == no_pyramid_axis:
+                shape.append(length)
+            else:
+                shape.append(max(1, length // 2))
+
+        # Stop if seen enough levels or level shape smaller than chunk size
+        if levels == -1:
+            if all(x <= c for x, c in zip(shape, chunk_size[-ndim:])):
+                break
+        elif level > levels:
+            break
+
+        print("Compute level", level, "with shape", shape)
+
+        allshapes.append(shape)
+        omz.create_dataset(str(level), shape=batch + shape, **opt)
+
+        # Iterate across `max_load` chunks
+        # (note that these are unrelated to underlying zarr chunks)
+        grid_shape = [ceildiv(n, max_load) for n in prev_shape]
+        for chunk_index in itertools.product(*[range(x) for x in grid_shape]):
+            print(f"chunk {chunk_index} / {tuple(grid_shape)})", end="\r")
+
+            # Read one chunk of data at the previous resolution
+            slicer = [Ellipsis] + [
+                slice(i * max_load, min((i + 1) * max_load, n))
+                for i, n in zip(chunk_index, prev_shape)
+            ]
+            fullshape = omz[str(level - 1)].shape
+            dat = omz[str(level - 1)][tuple(slicer)]
+
+            # Discard the last voxel along odd dimensions
+            crop = [
+                0 if y == 1 else x % 2 for x, y in zip(dat.shape[-ndim:], fullshape)
+            ]
```

Comment:
I might be mistaken but I think we need to slice the `fullshape` variable.

```suggestion
            crop = [
                0 if y == 1 else x % 2 for x, y in zip(dat.shape[-ndim:], fullshape[-ndim:])
            ]
```